[쪼밀리] 로또 1단계 미션 제출합니다.

chomily opened the pull request:
안녕하세요, 구구 리뷰어님
쪼밀리입니다!

아직 부족한게 많지만 열심히 미션 구현해보았습니다.
도움 주셔서 감사합니다 :)

chomily committed:
docs: README 작성

chomily committed:
feat: 로또 번호에 대한 검증

- 전달받은 인자가 1 - 45 이내인지 확인

pci2676 committed:
feat : 볼생성 equals 구현

 - override 이용하여 구현

pci2676 committed:
feat : 로또 티켓 구현 #1

chomily committed:
feat: LottoTicket equals 기능 구현

chomily committed:
feat: 로또 구매 금액 입력 기능 추가

pci2676 committed:
feat : 금액 입력 기능 작성

pci2676 committed:
feat : 로또티켓 발급기능 #1

chomily committed:
feat: 로또티켓 발급기능

pci2676 committed:
feat : 로또 티켓 번호 중복 체크 기능

chomily committed:
refactor: 로또팩토리, 티켓발급 클래스 분리

pci2676 committed:
feat : 로또 티켓발급 기능 테스트

chomily committed:
feat: 로또 우승 티켓 생성 기능 추가

pci2676 committed:
feat : 우승로또와 구매한 로또의 비교기능 작성

 - LottoResult 에 맞춘갯수, 보너스 맞춘 여부 포함

chomily committed:
feat: 로또 결과로 상금 확인

pci2676 committed:
refactor : 우승 판별 메서드 간소화

pci2676 committed:
feat : 우승 상금 통계화

chomily committed:
feat: 당첨 번호, 보너스 볼 입력 기능 추가

pci2676 committed:
refactor : 우승 상금 통계화 개선

pci2676 committed:
refactor : 로또 값 추가

chomily committed:
refactor: 결과 반환용 DTO 변경

- DTO 클래스 변경: Statistics, StatisticsDetail -> StatisticsDTO, PrizeDTO
- Statistics 내부에서 <K,V> PrizeGroup, StatisticsDetail을 이용한 정보 추출
   -> PrizeDTO 내부로 출력용 정보 이동

chomily committed:
feat: 로또 결과 출력 구현

chomily committed:
feat: 로또 게임 컨트롤러 생성

- 컨트롤러에서 사용할 BettingMoneyRequestDTO, WinningLottoRequestDTO 생성
- 입력오류 처리를 위한 ConvertFailException 생성
- LottoCompany에서 makeWinningLotto 메서드 매개변수 타입 변경(int[] -> Set<>)

pci2676 committed:
refactor : run 메소드 수정 및 일급컬렉션 작성

 - LottoTicket 의 컬렉션을 관리하는 LottoTicketBundle 작성
 - LottoResult 의 컬렉션을 관리하는 LottoResultBundle 작성
 - Controller 세부 구현사항을 Service 로 전달

pci2676 committed:
test : RequestDTO 누락된 테스트 추가

pci2676 committed:
refactor : DTO 에 Response 네이밍 추가

chomily committed:
refactor: 상수 추출, 컨벤션 적용, 테스트 코드 디렉토리 정리

chomily committed:
refactor: LottoCompany 추상화 및 구현체 추가

- 불필요한 메서드 제거
- 상수 추출
- 구조 변경에 따른 테스트 코드 추가 구현

chomily committed:
feat: 구매한 로또 번호 출력 기능

chomily committed:
fix: 입력값 trim 추가

chomily committed:
refactor: LottoBall 생성자 접근 제한 및 of 메서드 삭제

- 싱글톤으로 생성된 LottoBall을 LottoFactory를 통해서만 받을 수 있도록 수정
- LottoBall 받는 메서드명 변경 findLottoBallByNumber -> getLottoBallByNumber

chomily committed:
fix: LottoFactory.getLottoBallByNumber() 예외 처리 및 LottoBallValidator 분리

chomily committed:
refactor: LottoCompany -> LottoMachine, 추상 메서드 변경, DTO 구조 변경, 테스트 추가

- 네이밍 변경 LottoCompany -> LottoMachine
- LottoMachine 내부 추상 메서드 변경 buyTickets -> createOneTicket
- LottoTicketBundleResponseDTO 삭제 및 LottoTickResponseDTO 구조 변경

kang-hyungu commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,28 @@
+        this.lottoMachine = lottoMachine;
+    }
+
+    public LottoTicketBundle getLottoTicketBundle(BettingMoneyRequestDTO bettingMoneyRequestDTO) {
```
get 대신 create을 쓰면 어떨까요
객체를 생성하는 기능을 하고 있네요

chomily replied:
리뷰 주신 getXXX으로 네이밍된 메서드들 중 새로운 객체들을 생성해주는 메서드들은 createXXX로 네이밍을 변경하였습니다! &#128515;

수정하면서 다른 메서드들도 둘러보았는데,
컨트롤러 단에서 다른 메서드로부터 객체를 생성받아 바로 리턴하는 경우에는 getXXX으로 네이밍을 유지해줘도 괜찮을까요? 아니면 getter처럼 객체의 상태를 가져오는 경우에만 사용해주는게 좋을까요?

일반적으로 getXXX로 시작하는 메서드들은 getter로 동작할 것이라고 기대 되는 네이밍인가요?

``` java
    private LottoTicketBundle getLottoTicketBundle() {
        BettingMoneyRequestDTO bettingMoney = new BettingMoneyRequestDTO(InputView.inputBettingMoney());

        return service.createLottoTicketBundle(bettingMoney);
    }
```

kang-hyungu replied:
협업하는 다른 개발자가 내가 작성한 코드를 읽을 때를 생각해보면 의미 있는 메서드명을 쓰는게 좋습니다
모든 메서드가 get으로 시작하는 것보다 게임 규칙에 따른 이름으로 된 메서드명이 어떤 동작을 할지 빠르게 파악하는데 도움이 됩니다

kang-hyungu commented on src/main/java/lotto/domain/result/LottoResultBundle.java:
```diff
@@ -0,0 +1,39 @@
+        this.lottoResults = lottoResults;
+    }
+
+    public StatisticsResponseDTO getStatistics() {
```
도메인 클래스에서 dto를 반환하면 안됩니다
dto가 새로 생길 때마다 도메인 클래스에 계속 변경 사항이 생기게 됩니다.
서비스가 도메인을 사용해서 dto를 반환하게 수정해주세요

chomily replied:
네, 도메인에서 DTO 의존성을 모두 제거하였습니다 :)

kang-hyungu commented on src/main/java/lotto/domain/result/LottoResultBundle.java:
```diff
@@ -0,0 +1,39 @@
+public class LottoResultBundle {
+    private final List<LottoResult> lottoResults;
+
+    public LottoResultBundle(List<LottoResult> lottoResults) {
```
null 일 때 어떻게 처리할지도 추가하는게 좋겠네요

chomily replied:
null이 전달될 때, Exception 발생하도록 유효성 검사 추가하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/result/LottoResultBundle.java:
```diff
@@ -0,0 +1,39 @@
+        return new StatisticsResponseDTO(getPrizeGroup());
+    }
+
+    private List<PrizeGroup> getPrizeGroup() {
```
get 대신 다른 메소드명을 써보면 어떨까요?
get을 쓰면 getter처럼 보이는데 실제 동작은 stream 써서 PrizeGroup로 변환 처리를 하네요

chomily replied:
리팩토링으로 해당 메서드가 삭제 되었습니다. 그래서 기존 getPrizeGroup()와 유사한 역할을 하는 메서드의 명을 createLottoMatchResult()로 적용하고 내부에서 비슷한 동작을 하는 메서드들도 create를 이용하여 네이밍해보았습니다!

- createLottoMatchResult
- createLottoMatchResultBundle

kang-hyungu commented on src/main/java/lotto/domain/result/win/WinningBalls.java:
```diff
@@ -0,0 +1,36 @@
+
+    public WinningBalls(Set<LottoBall> lottoBalls, LottoBall bonusBall) {
+        this.lottoBalls = new HashSet<>(lottoBalls);
+        this.lottoBalls.add(bonusBall);
```
bonusBall가 의도한 값이 안들어올 때 처리도 필요해보입니다

chomily replied:
WinLottoTicket클래스에서 당첨 번호와 Bonus 번호의 중복을 검출 할 수 있도록 로직을 추가하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/result/win/WinningBalls.java:
```diff
@@ -0,0 +1,36 @@
+        this.lottoBalls.add(bonusBall);
+    }
+
+    public int getMatchCount(LottoTicket lottoTicket) {
```
getter가 아니네요
메서드명을 의미가 더 잘 드러나게 바꿔볼까요?

chomily replied:
실제로 사용하는 부분에서 읽어보니 메서드명에서 어색함을 느낄 수 있었습니다.
그래서 일치하는 번호를 확인한다는 의미로 메서드 명을 countMatchNumber으로 변경하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/result/win/WinningLotto.java:
```diff
@@ -0,0 +1,39 @@
+    private final WinningBalls winningBalls;
+    private final LottoBall bonusBall;
+
+    public WinningLotto(Set<LottoBall> lottoBalls, LottoBall bonusBall) {
```
이 생성자도 인자들이 null일 때 처리가 필요해보이네요

chomily replied:
null이 전달될 때, Exception 발생하도록 유효성 검사 추가하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/result/win/WinningLotto.java:
```diff
@@ -0,0 +1,39 @@
+        this.bonusBall = bonusBall;
+    }
+
+    public LottoResult getResult(LottoTicket buyLottoTicket) {
```
객체를 생성하는 메서드네요
create이 어울리지 않을까요?

chomily replied:
네! createLottoResult로 변경완료하였습니다~

kang-hyungu commented on src/main/java/lotto/domain/result/win/prize/PrizeGroup.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private boolean isThird(LottoResult lottoResult) {
+        return this == SECOND && !lottoResult.isBonusMatch();
```
values라 순서가 보장되긴 할테지만 안전하게 (this == SECOND || this == THIRD) 로 하면 어떨까요?

chomily replied:
현재는 순서가 보장되는데 Enum 순서나 코드가 변경될 때를 대비해서 둘 다 적어 주는 것이 안전하고 좋을 것 같아요! 반영하였습니다 :)

kang-hyungu commented on src/main/java/lotto/domain/result/win/prize/PrizeGroup.java:
```diff
@@ -0,0 +1,52 @@
+        this.defaultPrize = defaultPrize;
+    }
+
+    public static PrizeGroup findPrizeByLottoResult(LottoResult result) {
```
result가 null이면 몇등일까요?

chomily replied:
null이 전달될 때, Exception 발생하도록 유효성 검사 추가하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/result/win/prize/PrizeGroup.java:
```diff
@@ -0,0 +1,52 @@
+        return this == SECOND && !lottoResult.isBonusMatch();
+    }
+
+    public static StatisticsResponseDTO toDtos(List<PrizeGroup> prizeGroups) {
```
도메인 클래스에서 dto를 생성하면 안됩니다
서비스 계층으로 옮겨봐요

chomily replied:
네, 도메인에서 DTO 의존성을 모두 제거하였습니다 :)

kang-hyungu commented on src/main/java/lotto/domain/ticket/LottoMachine.java:
```diff
@@ -0,0 +1,38 @@
+
+import static lotto.domain.ticket.LottoTicket.LOTTO_PRICE;
+
+public abstract class LottoMachine {
```
메서드들이 인자로 dto를 받고 있네요
dto에서 쓰려는 값들만 인자로 받으면 되지 않을까요?
인자로 dto를 받아오면 도메인이 서비스 계층에 의존하는 코드가 됩니다

chomily replied:
서비스에서 필요한 값으로 변환시켜 인자로 받도록 수정하였습니다 :)

kang-hyungu commented on src/main/java/lotto/domain/ticket/LottoTicket.java:
```diff
@@ -0,0 +1,55 @@
+
+    private final Set<LottoBall> lottoBalls;
+
+    public LottoTicket(Set<LottoBall> lottoBalls) {
```
null 체크가 필요합니다

chomily replied:
null이 전달될 때, Exception 발생하도록 유효성 검사 추가하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/ticket/LottoTicketBundle.java:
```diff
@@ -0,0 +1,40 @@
+public class LottoTicketBundle {
+    private final List<LottoTicket> lottoTickets;
+
+    public LottoTicketBundle(List<LottoTicket> lottoTickets) {
```
null 체크가 필요합니다

chomily replied:
null이 전달될 때, Exception 발생하도록 유효성 검사 추가하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/ticket/LottoTicketBundle.java:
```diff
@@ -0,0 +1,40 @@
+        this.lottoTickets = lottoTickets;
+    }
+
+    public LottoResultBundle getLottoResults(WinningLotto winningLotto) {
```
당첨 로또로 티켓 결과를 새로 받아오는데 get이 맞을까요?

chomily replied:
위 메서드도 createLottoResultBundle 이름 변경하였습니다!

kang-hyungu commented on src/main/java/lotto/domain/ticket/ball/LottoBallValidator.java:
```diff
@@ -0,0 +1,13 @@
+package lotto.domain.ticket.ball;
+
+public class LottoBallValidator {
+    public static final int LOWER_BOUND = 1;
```
검증하는 클래스가 갖고 있기 보단 실제 게임 규칙에 관여하는 클래스가 갖는게 좋지 않을까요?
로또 번호 범위가 어떻게 되는지 확인하려면 검증 클래스가 아닌 도메인 클래스를 먼저 확인하지 않을까요?

chomily replied:
LottoBall과 LottoBallFactory에서 중복코드가 발생하여 클래스로 분리했었는데, 로또 볼에 관한 규칙은 로또 볼에서 찾아보려 할 것 같습니다!
LottoBall으로 검증 로직을 이동시키고, 해당 메서드를 public으로 열어 Factory에서도 같이 사용할 수 있도록 수정해보았습니다.

kang-hyungu commented on src/main/java/lotto/view/dto/LottoTicketResponseDTO.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Integer[] getNumbers() {
+        return this.numbers.toArray(new Integer[6]);
```
매직넘버가 있네요

chomily replied:
앗, 매직넘버 상수로 추출하였습니다!

kang-hyungu commented on src/main/java/lotto/view/dto/PrizeResponseDTO.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public double getSum() {
+        return defaultPrize * matchTicketCount;
```
게임 규칙이 dto에 있으면 안됩니다
도메인 계층에서 처리하고 dto로 넘겨주는게 좋겠네요

chomily replied:
넵, DTO내부 도메인 관련 로직 모두 추출였습니당

kang-hyungu commented on src/main/java/lotto/view/dto/StatisticsResponseDTO.java:
```diff
@@ -0,0 +1,58 @@
+
+import static lotto.domain.ticket.LottoTicket.LOTTO_PRICE;
+
+public class StatisticsResponseDTO {
```
dto 코드에 도메인 코드에 관련된 로직이 있으면 안됩니다
단순하게 값을 전달 받고 반환해줘야 합니다
도메인 클래스로 추출하는게 좋겠네요

chomily replied:
DTO내부 도메인 관련 로직 모두 추출였습니당

kang-hyungu commented on src/main/java/lotto/view/dto/WinningLottoRequestDTO.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class WinningLottoRequestDTO {
```
이 클래스도 마찬가지로 dto 이상의 책임을 갖고 있네요
뷰와 도메인 클래스에게 책임을 나눠주세요

chomily replied:
DTO내부 도메인 관련 로직 제거하고 뷰와 도메인으로 책임 이동하였습니다

kang-hyungu reviewed (changes requested):
안녕하세요 쪼밀리! 리뷰어 구구입니다 👋 
객체 설계 및 계층 분리가 잘 되어 있네요 👍 
피드백 남긴 부분 확인해보시고 궁금한 점 생기면 코멘트나 dm 남겨주세요

chomily committed:
test: 누락 테스트 추가

chomily committed:
refactor: 불명확한 네이밍 수정 및 Controller 구조 변경

- DTO의 Response, Request 네이밍 삭제
- PrizeGroup -> Rank
- getXXX 메서드들 네이밍 변경

chomily committed:
refactor: WinLottoTicket 구조 및 패키지 수정

- 클래스 네이밍 변경 WinningLotto -> WinLottoTicket
- WinLottoTicket 의 멤버 WinningBalls 를 LottoTicket 으로 변경
- 불필요한 WinningBalls 클래스 삭제
- WinLottoTicket 패키지 이동(result/win -> ticket)
- 불필요한 패키지 단계 삭제(win)

chomily committed:
fix: 로또 당첨 번호와 보너스볼 중복 검사

chomily committed:
refactor: 메서드 네이밍 수정 및 불필요한 메서드 삭제

chomily committed:
refactor: 해당 Rank 확인 조건 추가

chomily committed:
refactor: 도메인의 BettingMoneyDTO 의존성 제거

- BettingMoneyDTO 내부의 도메인 로직 LottoMachine 으로 이동
- 입력 금액에 대한 예외 테스트 LottoMachineTest로 이동
- BettingMoneyDTOTest 삭제

chomily committed:
refactor: LottoTicketDTO 생성 위치 변경, 메직 넘버 제거, 불필요한 테스트 삭제

chomily committed:
refactor: 도메인의 WinLottoTicketDTO 의존성 제거

- WinLottoTicketDTO 구조 변경 (멤버 Set -> List)
- 도메인에 해당하는 유효성 검사 제거
- 테스트 이동

chomily committed:
refactor: LottoResult 구조 변경, ResultDTO의 도메인 로직 추출 및 RankDTO 삭제

- (추후 리팩토링 예정) 전체 수익률 등 계산을 위한 OverallResult 클래스 작성

chomily committed:
refactor: 네이밍 수정 및 OverallResult 분리

- 네이밍 수정 LottoResult, LottResultBundle -> LottoMatchResult, LottoMatchResultBundle
- 출력용 DTO생성 위치 이둥
- 서비스 내 도메인 로직 OverallResult로 책임 전달

chomily committed:
refactor: 로또 볼 검증 로직 도메인으로 이동, LottoBallValidator 삭제

chomily committed:
refactor: NullOrEmptyValidator 작성

- 객체를 매개변수로 받는 생성자, 메서드 null 체크 추가

kang-hyungu commented on src/main/java/lotto/util/NullOrEmptyValidator.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Map;
+
+public class NullOrEmptyValidator {
+    public static void isNull(Object object) {
```
자바에서 기본적으로 제공하는 Objects.isNull()을 쓰면 이 메서드는 없어도 될거에요

chomily replied:
감사합니다!
`object == null` 대신 `Objects.isNull()`을 사용도록 수정해볼게요!
여러 곳에서 null체크를 하는데, 해당 조건이 true일 경우 excpetion발생하는 라인이 반복적으로 들어가게 되어 메서드로 빼두었는데 어떻게 생각하시나요?!

kang-hyungu replied:
이건 취향 문제일수도 있겠네요
저는 공통으로 빼진 않고 각 생성자마다 if절을 써주는 편이에요
예외 대신 다른 처리를 해줘야하는 경우도 있더라구요
에러 메시지를 다르게 할수도 있구요
편한 방법으로 사용하시면 될 것 같아요!

kang-hyungu commented on src/main/java/lotto/view/dto/LottoTicketDTO.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Collectors;
+
+public class LottoTicketDTO {
+    private static final int NUMBER_OF_LOTTO_BALLS = 6;
```
이 값은 도메인에 있어야겠네요

chomily replied:
상수값 6을 사용하지 않도록 로직을 수정하였습니다!

kang-hyungu commented on src/main/java/lotto/view/dto/LottoTicketDTO.java:
```diff
@@ -0,0 +1,23 @@
+public class LottoTicketDTO {
+    private static final int NUMBER_OF_LOTTO_BALLS = 6;
+
+    private final Integer[] numbers;
```
배열 대신 List를 써볼까요?

chomily replied:
List로 변경하였어요

kang-hyungu commented on src/main/java/lotto/view/dto/WinLottoTicketDTO.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.stream.Collectors;
+
+public class WinLottoTicketDTO {
+    private static final String PARSE_FAIL_EXCEPTION_MESSAGE = "%s : 숫자가 아닌 문자가 존재합니다.";
```
dto는 단순 값만 전달해야 합니다
이 상수들은 view로 옮기는게 좋겠어요

chomily replied:
네, 상수와 파싱 로직 같이 InputView로 이동하였습니다

kang-hyungu commented on src/main/java/lotto/view/dto/WinLottoTicketDTO.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private List<Integer> splitAndConvertToInteger(String winningNumbers) {
```
파싱하는 작업도 dto가 아닌 view에서 처리해도 되겠네요

chomily replied:
앗 배열로 파싱하는 작업도 InputView의 역할이네요! 고맙습니다:) 로직 이동하였습니다

kang-hyungu commented on src/test/java/lotto/domain/ticket/LottoMachineForTest.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class LottoMachineForTest extends LottoMachine {
```
테스트 코드를 production 코드 상속 받아서 구현하면 안됩니다
LottoMachine를 객체로 구현해서 테스트 진행하는게 좋습니다

kang-hyungu reviewed (approved):
피드백 반영 잘 하셨어요 👍 
간단히 몇 개 남긴 피드백은 다음 미션 진행하면서 반영해보아요
pr은 머지 처리할게요 수고하셨어요!
